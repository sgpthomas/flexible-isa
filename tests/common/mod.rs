use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
