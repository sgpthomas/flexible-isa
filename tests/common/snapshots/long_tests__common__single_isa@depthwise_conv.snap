---
source: tests/common/mod.rs
description: generate isa for depthwise_conv
---
[
    (
        47,
        1,
        "(+ (cast<int32x128> (/ (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> ?#3))))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?x1))) (cast<int32> ?#5))) (cast<int32> ?x2))))))",
    ),
    (
        124,
        1,
        "(+ (cast<int32x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#6 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        128,
        1,
        "(+ (cast<int32x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> 384) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#6 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        154,
        1,
        "(- (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        174,
        1,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> ?x1))",
    ),
    (
        192,
        1,
        "(< (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        209,
        1,
        "(if (cast<uint8x128> (get ?x0 (cast<int32x128> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 128))))) (cast<bool> ?x2))",
    ),
    (
        214,
        1,
        "(if ?x0 (cast<bool> ?x1))",
    ),
    (
        216,
        1,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        217,
        1,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        7,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> ?#3))) (cast<int32> ?x0))",
    ),
    (
        8,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> ?#3))) (cast<int32> ?x0))) (cast<int32> ?#4))) (cast<int32> ?x1))",
    ),
    (
        11,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?#2))) (cast<int32> ?#3))) (cast<int32> (- (cast<int32> ?#4) (cast<int32> ?x0))))",
    ),
    (
        26,
        2,
        "(+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> 0) (cast<int32> ?x1))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> ?x3) (cast<int32> ?x4))) (cast<int32> ?x5))) (cast<int32> ?x6))))",
    ),
    (
        44,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> ?x1))))",
    ),
    (
        126,
        2,
        "(+ (cast<int32x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#6 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        160,
        2,
        "(- (cast<int32> (call[max] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?#18) (cast<int32> ?x0) (cast<int32> (call[select] (cast<bool> ?x1) (cast<int32> 0) (cast<int32> ?x2))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x3) (cast<int32> ?x4) (cast<int32> ?x5))) (cast<int32> ?x6))) (cast<int32> ?x7))))) (cast<int32> ?x8))) (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?#19) (cast<int32> ?x9) (cast<int32> (call[select] (cast<bool> ?x1) (cast<int32> ?x2) (cast<int32> 0))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x3) (cast<int32> ?x5) (cast<int32> ?x4))) (cast<int32> ?x6))) (cast<int32> ?x7))))) (cast<int32> ?x10))))",
    ),
    (
        167,
        2,
        "(* (cast<int32> (+ (cast<int32> (- (cast<int32> ?#2) (cast<int32> ?x0))) (cast<int32> ?x1))) (cast<int32> ?#3))",
    ),
    (
        173,
        2,
        "(* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> ?x4))",
    ),
    (
        180,
        2,
        "(* (cast<int32> ?x0) (cast<int32> 2))",
    ),
    (
        185,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> 2))",
    ),
    (
        186,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        194,
        2,
        "(<= (cast<int16x64> (call[int16x64] (cast<int32x64> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 64))))) (cast<int16x64> (call[x64] (cast<int16> (call[saturating_cast] (cast<int32> (call[saturating_sub] (cast<int32> ?#1) (cast<int32> ?x0))))))))",
    ),
    (
        206,
        2,
        "(|| (cast<bool> (<= (cast<int32> 8) (cast<int32> ?x0))) (cast<bool> (== (cast<int32> (\"%\" (cast<int32> ?x0) (cast<int32> 2))) (cast<int32> 0))))",
    ),
    (
        211,
        2,
        "(if (cast<uint8x128> (get ?#1 (cast<int32x128> ?x0))) (cast<bool> (<= (cast<int16x128> (call[int16x128] (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> (call[saturating_cast] (cast<int32> (call[saturating_sub] (cast<int32> ?#2) (cast<int32> 1))))))))))",
    ),
    (
        212,
        2,
        "(if (cast<uint8x64> (get ?#3 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 64))))) (cast<bool> (<= (cast<int8x64> (call[int8x64] (cast<int32x64> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 64))))) (cast<int8x64> (call[x64] (cast<int8> (call[saturating_cast] (cast<int32> (call[saturating_sub] (cast<int32> ?#4) (cast<int32> ?x2))))))))))",
    ),
    (
        237,
        2,
        "(get ?#1 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> ?#3))) (cast<int32> ?x0))) (cast<int32> 64))) (cast<int32> 1) (cast<int32> 64))))",
    ),
    (
        239,
        2,
        "(get ?#2 (cast<int32x128> (+ (cast<int32x128> (/ (cast<int32x128> (call[ramp] (cast<int32> ?#3) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> ?#4))))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?x0))) (cast<int32> ?#6))) (cast<int32> ?x1))))))))",
    ),
    (
        240,
        2,
        "(get ?#2 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#3) (cast<int32> ?x0))) (cast<int32> ?#4))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        10,
        3,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 64))",
    ),
    (
        12,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 128))",
    ),
    (
        13,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 256))",
    ),
    (
        15,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 128))",
    ),
    (
        16,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 256))",
    ),
    (
        48,
        3,
        "(+ (cast<int32x128> (get ?#1 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))) (cast<int32> ?#6))) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        49,
        3,
        "(+ (cast<int32x128> (get ?#1 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))) (cast<int32> ?#6))) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        50,
        3,
        "(+ (cast<int32x128> (get ?#1 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?#5))) (cast<int32> ?#6))) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        51,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        52,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        53,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        54,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        59,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> ?#4) (cast<int32> ?x0))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#5 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> ?#6) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        60,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        61,
        3,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> ?x0) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        162,
        3,
        "(- (cast<uint8x128> ?x0) (cast<int32x128> (call[widen_right_mul] (cast<int32x128> ?x1) (cast<int16x128> (call[x128] (cast<int16> (call[int16] (cast<uint8> ?#2))))))))",
    ),
    (
        182,
        3,
        "(* (cast<int32> ?x0) (cast<int32> 256))",
    ),
    (
        201,
        3,
        "(== (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        208,
        3,
        "(if (cast<int32x128> (get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))) (cast<bool> ?x1))",
    ),
    (
        219,
        3,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        222,
        3,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> ?x1) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        238,
        3,
        "(get ?#1 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> ?#3))) (cast<int32> ?x0))) (cast<int32> 1) (cast<int32> 64))))",
    ),
    (
        25,
        4,
        "(+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> 0) (cast<int32> ?x1))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> 0) (cast<int32> (+ (cast<int32> ?x3) (cast<int32> (neg (cast<int32> 1))))))) (cast<int32> ?x4))) (cast<int32> ?x5))))",
    ),
    (
        42,
        4,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        155,
        4,
        "(- (cast<int32> (call[max] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> 0) (cast<int32> ?x3))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x4) (cast<int32> 0) (cast<int32> (+ (cast<int32> ?x5) (cast<int32> (neg (cast<int32> 1))))))) (cast<int32> ?x6))) (cast<int32> ?x7))))) (cast<int32> ?x8))) (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x9) (cast<int32> (call[select] (cast<bool> ?x2) (cast<int32> ?x3) (cast<int32> 0))))) (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x4) (cast<int32> ?x5) (cast<int32> 1))) (cast<int32> ?x6))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?x7))))) (cast<int32> ?x10))))",
    ),
    (
        164,
        4,
        "(* (cast<int32> (+ (cast<int32> (+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> 1))) (cast<int32> ?x2))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?x3))",
    ),
    (
        181,
        4,
        "(* (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        196,
        4,
        "(<= (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        20,
        6,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> (+ (cast<int32> (* (cast<int32> ?#5) (cast<int32> ?x2))) (cast<int32> (* (cast<int32> ?x3) (cast<int32> ?x4))))))",
    ),
    (
        38,
        6,
        "(+ (cast<int32> (call[select] (cast<bool> ?x0) (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        43,
        6,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 2))))",
    ),
    (
        153,
        6,
        "(- (cast<int32> (* (cast<int32> ?x0) (cast<int32> 2))) (cast<int32> ?x1))",
    ),
    (
        175,
        6,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?x1))",
    ),
    (
        178,
        6,
        "(* (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 0))) (cast<int32> 2))",
    ),
    (
        179,
        6,
        "(* (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> 0))) (cast<int32> 2))",
    ),
    (
        22,
        7,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        152,
        7,
        "(- (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        204,
        7,
        "(&& (cast<bool> ?x0) (cast<bool> ?x1))",
    ),
    (
        75,
        8,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 256) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        177,
        8,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        69,
        9,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 128) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> 1))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        81,
        9,
        "(+ (cast<int32x128> (get ?#2 (cast<int32x128> (call[ramp] (cast<int32> 384) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widening_mul] (cast<int16x128> (get ?#3 (cast<int32x128> (call[ramp] (cast<int32> ?x0) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#4 (cast<int32x128> (call[ramp] (cast<int32> (* (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#5) (cast<int32> ?#6))) (cast<int32> 1))) (cast<int32> ?#7))) (cast<int32> ?x1))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        198,
        9,
        "(<= ?x0 ?x1)",
    ),
    (
        176,
        10,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 128))",
    ),
    (
        200,
        10,
        "(== (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        191,
        12,
        "(< (cast<int32> ?x0) (cast<int32> 0))",
    ),
    (
        215,
        12,
        "(get ?x0 (cast<int32x128> (call[ramp] (cast<int32> 0) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        9,
        19,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        40,
        19,
        "(+ (cast<int32> ?x0) (cast<int32> 1))",
    ),
    (
        183,
        20,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        46,
        42,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        161,
        43,
        "(- (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
]
